File: examples/enterprise/custom_llm.py
"""
Custom LLM Integration (Anthropic Claude)
==========================================
Use your own LLM instead of OpenAI Realtime. Patter handles STT, TTS,
and telephony while you control the brain. This example uses Anthropic
Claude via the Messages API, but any LLM works — OpenAI, Mistral,
LLaMA, or your own fine-tuned model.

Requirements:
    pip install patter python-dotenv httpx

Environment variables (.env):
    ANTHROPIC_API_KEY    - Your Anthropic API key
    DEEPGRAM_API_KEY     - Deepgram key for speech-to-text
    ELEVENLABS_API_KEY   - ElevenLabs key for text-to-speech
    TWILIO_ACCOUNT_SID   - Twilio account SID
    TWILIO_AUTH_TOKEN    - Twilio auth token
    TWILIO_PHONE_NUMBER  - Your Twilio phone number (E.164 format)
    WEBHOOK_URL          - Public URL where Twilio can reach this server
"""

import asyncio
import os

import httpx
from dotenv import load_dotenv

load_dotenv()

from patter import Patter

ANTHROPIC_API_KEY = os.environ["ANTHROPIC_API_KEY"]
SYSTEM_PROMPT = (
    "You are a concise, friendly phone assistant for Acme Corp. "
    "Help callers with account questions, billing, and general inquiries. "
    "Keep responses under two sentences — this is a phone call, not a chat."
)


async def on_message(data: dict) -> str:
    """Called each time the caller finishes speaking.

    Patter transcribes speech via Deepgram (STT), passes the text here,
    and synthesises your return value via ElevenLabs (TTS).

    Args:
        data: {"text": str, "history": [{"role": "user"|"assistant", "text": str}],
               "call_id": str, "caller": str}
    """
    history = data["history"]

    # Build the messages payload for Claude
    messages = [{"role": m["role"], "content": m["text"]} for m in history]

    async with httpx.AsyncClient(timeout=15.0) as client:
        response = await client.post(
            "https://api.anthropic.com/v1/messages",
            headers={
                "x-api-key": ANTHROPIC_API_KEY,
                "anthropic-version": "2023-06-01",
                "content-type": "application/json",
            },
            json={
                "model": "claude-sonnet-4-20250514",
                "max_tokens": 256,
                "system": SYSTEM_PROMPT,
                "messages": messages,
            },
        )
        response.raise_for_status()
        result = response.json()

    return result["content"][0]["text"]


# ── Initialise Patter in pipeline mode ───────────────────────────────
phone = Patter(
    mode="local",
    twilio_sid=os.getenv("TWILIO_ACCOUNT_SID"),
    twilio_token=os.getenv("TWILIO_AUTH_TOKEN"),
    phone_number=os.getenv("TWILIO_PHONE_NUMBER"),
    webhook_url=os.getenv("WEBHOOK_URL"),
)

# Pipeline mode: Patter does STT + TTS, you provide the LLM
agent = phone.agent(
    provider="pipeline",
    system_prompt=SYSTEM_PROMPT,
    stt=Patter.deepgram(api_key=os.environ["DEEPGRAM_API_KEY"]),
    tts=Patter.elevenlabs(api_key=os.environ["ELEVENLABS_API_KEY"], voice="aria"),
    language="en",
)

# ── Start the server ─────────────────────────────────────────────────
if __name__ == "__main__":
    print("Listening for calls (pipeline mode with Claude)...")
    asyncio.run(phone.serve(agent, port=8000, on_message=on_message))

File: examples/enterprise/custom_llm.ts
/**
 * @file custom_llm.ts
 * @description Bring your own LLM with Patter's pipeline mode.
 *
 * Patter handles speech-to-text and text-to-speech while you control the
 * "brain" — any LLM, fine-tuned model, or deterministic logic. This example
 * uses Anthropic Claude via a simple fetch call, but the pattern works with
 * any HTTP-accessible model (OpenAI, Mistral, Llama, etc.).
 *
 * How it works:
 *   1. Caller speaks -> Patter transcribes via Deepgram STT
 *   2. Transcript arrives in your onMessage handler
 *   3. You call your own LLM and return the response string
 *   4. Patter speaks the response via ElevenLabs TTS
 *
 * Prerequisites:
 *   - ANTHROPIC_API_KEY in .env (or whichever LLM you use)
 *   - TWILIO_ACCOUNT_SID, TWILIO_AUTH_TOKEN, TWILIO_PHONE_NUMBER in .env
 *   - DEEPGRAM_API_KEY, ELEVENLABS_API_KEY in .env
 *   - WEBHOOK_URL pointing to this server (e.g. ngrok)
 *
 * Usage:
 *   npx ts-node custom_llm.ts
 */

import { Patter, deepgram, elevenlabs } from "patter";
import dotenv from "dotenv";

dotenv.config();

const phone = new Patter({
  mode: "local",
  twilioSid: process.env.TWILIO_ACCOUNT_SID!,
  twilioToken: process.env.TWILIO_AUTH_TOKEN!,
  phoneNumber: process.env.TWILIO_PHONE_NUMBER!,
  webhookUrl: process.env.WEBHOOK_URL!,
});

const agent = phone.agent({
  systemPrompt: "You are a helpful enterprise support agent.",
  provider: "pipeline",
  stt: deepgram({ apiKey: process.env.DEEPGRAM_API_KEY! }),
  tts: elevenlabs({ apiKey: process.env.ELEVENLABS_API_KEY!, voice: "rachel" }),
  firstMessage: "Hello, how can I assist you today?",
});

/** Call your own LLM — replace with any model or routing logic. */
async function callAnthropic(userMessage: string): Promise<string> {
  const response = await fetch("https://api.anthropic.com/v1/messages", {
    method: "POST",
    headers: {
      "x-api-key": process.env.ANTHROPIC_API_KEY!,
      "anthropic-version": "2023-06-01",
      "content-type": "application/json",
    },
    body: JSON.stringify({
      model: "claude-sonnet-4-20250514",
      max_tokens: 256,
      system: agent.systemPrompt,
      messages: [{ role: "user", content: userMessage }],
    }),
  });

  const body = (await response.json()) as Record<string, unknown>;
  const content = body.content as Array<{ text: string }>;
  return content[0].text;
}

async function main(): Promise<void> {
  await phone.serve({
    agent,
    port: 8000,
    onMessage: async (data) => {
      const transcript = data.transcript as string;
      console.log(`User said: ${transcript}`);

      const reply = await callAnthropic(transcript);
      console.log(`LLM replied: ${reply}`);
      return reply;
    },
  });

  console.log("Custom LLM agent listening on port 8000");
}

main();
